docs/secrets.md

```diff
@@ -0,0 +1,44 @@
+Notification endpoints will require credentials in order to communicate with the various external services.
+
+We do not want users to include those credentials in plain text into Flux scripts. We need to provide a mechanism for users load secrets from a safe storage mechanism and provide those secrets to the endpoints.
+
+**Backing Store:** : There are many backing store possibilities, which need not be discussed here.  Any store that provides secure access for key/value lookup may be used by providing an implementation of the KeyLookup  interface in the flux repository.  
+
+**Fine Grained Authorization**:  keys will generally not be accessible in the flux language.  A secret-enabled function will accept meta-data about the secret (a key) and use it to look up the stored value in memory. The lookup will error if the key is not found or if the user is not authorized to use that key.  
+
+**Design**
+
+A function that requires secret lookup must: 
+
+- Be built-in to the language
+- Register an ExternalDependencies struct  (note: it will be straightforward and advantageous to refactor the `InjectFromDependencies` and similar functions into this pattern as well).   
+- Not return secrets under any circumstances (e.g. in error messages) 
+
+The external interface for secret lookup is: 
+```Go
+func LookupKey(keyObject values.Object) (string, error)
+```
+The keyObject is a values.Object type because: 
+
+- The secret parameter can be passed to the service without being modified
+- The parameter may be validated by the service, minimizing the dependency of the calling function to have prior knowledge of the key lookup protocol. 
+
+The keyObject should minimally have a `secretKey` property: 
+slackID = {secretKey: "FLUX_SLACK_ID"}
+
+but the object may include supplementary information for consumption by the calling function.  They may also be embedded + replaced in another values.Value type (e.g. an encapsulating Object or Array) by the calling function as in:
+
+http.post(header: {"Content Type": "application-json", "ApplicationSecret": {secretKey: "FLUX_HTTP_POST_KEY"}}, url: "http://some/url/", data: json_enc)
+
+
+A final implementation may look like: 
+```Go
+// calling flux:  slack.message(message: {token: {secretKey: "FLUX_SLACK_ID"}, text: "hello slack"})
```

Comment:
I think it could be more user friendly to  have this resulting flux:

```
import "vault"
token = vault.get("mytoken")
slack.message(token: token, text: "hello slack")
```

We could achieve this by defining a `secret` type that is basically a promise for a `string`, so that in Flux you won't be able of calling it, accessing its parts, print it, etc.

In Go, you can use its internal function and retrieve the token.

Reply:
I spent some time discussing with Nathaniel about this. we determined not to implement a new type.    However, my design is compatible with this look/feel.  

`vault.Get(key: "SLACK_ID")`  can be provided to create the object so that users don't have to memorize/risk mistyping the _required_  object type schema.  This would also allow us to extend the API in a backward compatible way.  

I do kind of like the idea of a token being a promise for a string, but it would open up security risks:  what functions are privileged to evaluate the `token` as a string?  Can it be evaluated via string interpolation (a forthcoming feature)?  When is it evaluated?  do we delay string interpolation until runtime? If they do evaluate the token in a non-privileged setting, do we error or provide a description of the promise?  So then the identifier has a different evaluation depending on context, which is unusual.   it's reasons like these that we ruled out "it's a string, but not really"  type of approaches.  

Also considered:  encrypting the keyvalue and just storing it as a regular string.  this also has complexity in terms of key management and proper security protocols to prevent a breach.

Reply:
I see your point.

Anyways, will that object be "special". I mean nobody should get anything out from it. At least not in  Flux.

The Go code can extract the secret. Wasn't this the assumption?

I am  quite not following how that object masks the token from the doc 🤔

Reply:
it's just   

`vault.Get = (key) = ({secretKey: key})`

or if we determine that the API needs more information (userid, orgid, ...)  we can add it there too.

Reply:
Yeah, I got it. I think we are on the same page. It's not technically a promise for the token, but everything  that Go needs to build a token.

I wonder if we could obtain the token and keep it isolated in the Flux context.

What I mean is that I expect that:

```
token = vault.get("SLACK_ID")
```

would  error if there is some connection problem or else.

I am not expecting lazy evaluation here 🤔 .

Anyways, I agree that this is a valid  design :)